### ISSUE TITLE
recipe for hyperparameter sweep

### ISSUE THREAD

**Issue by felipemello1** [pytorch/torchtune#1752]
Torchtune could provide a recipe to do HPO, where the user provides a config, the recipe, eval dataset, params to sweep and budget.

I just played with optimizer. Our default in lr 3e-4. I tried 3e-3 and the ScheduleFree optimizer. Big difference.

![image](https://github.com/user-attachments/assets/4bde829a-a45c-4a43-8cd3-2d5910d2aff2)

I don't know how well it works for finetuning, but an extension could be to allow HPO with smaller models, and transfer the best settings to the large models using muP (https://github.com/microsoft/mup).

**Comment by RdoubleA** [comment]
We would need some form of validation during training before we can do HPO, no?

On a related note, can we even run eval on a custom benchmark dataset that's not in EleutherAI?

**Comment by felipemello1** [comment]
I don’t know if necessarily during training. Maybe it’s ok to train for X steps and then do eval. I also think that as long as you are training for a single epoch, the train loss may be a good enough proxy, since the data doesn’t get repeated, and therefore there is no memorization

**Comment by krammnic** [comment]
Isn't it actually require some changes in optimizer, model, basic parameters if doing muP?

**Comment by krammnic** [comment]
Make it naturally with just "Provide recipe" might be pretty non-trivial

**Comment by ebsmothers** [comment]
Just catching up on this issue. Tbh I think this is a nontrivial digression from what we would traditionally consider a recipe. What I've seen in the past is that sweep functionality is usually provided as its own script wrapping the CLI, I don't think the concept of a recipe would apply in this case. 

Also I would distinguish between hyperparameter optimization vs sweeping. Maybe this is just a matter of semantics, but in my mind hyperparameter optimization implies we are taking an active role in tuning the parameters to maximize the expectation of some conditional distribution (i.e. whatever validation metric we're optimizing for). That is a lot more work than basic sweep functionality. Though even doing sweeps properly requires a bit of accounting (e.g. do we execute multiple runs in parallel, how do we maintain artifacts, if we have multiple distributed runs at once what ports do we use, etc.).

So yeah I guess imo this is not something we should be tackling right now.

**Comment by felipemello1** [comment]
> So yeah I guess imo this is not something we should be tackling right now.

100% agree its not a priority! However, sweeping at least the LR is a big deal, and the way I see torchtune is that we have a training library with pre-filled recipes that are ready to go, even though people can make their own. My understanding is that every time we provide a recipe, more people use torchtune because its convenient. But i wanted to leave the idea here just as a placeholder, and not as a "this should be our next big step".

**Comment by krammnic** [comment]
@ebsmothers @felipemello1 As compromise I think it will be good if just there will be some API maybe arguments in recipes that will help user do muP without touching internals of tuning process(through recipe)
